exoscale/datasource_exoscale_elastic_ip.go

```diff
@@ -125,29 +128,73 @@ func dataSourceElasticIPRead(ctx context.Context, d *schema.ResourceData, meta i
 
 	client := GetComputeClient(meta)
 
-	elasticIPID, byElasticIPID := d.GetOk(dsElasticIPAttrID)
-	elasticIPAddress, byElasticIPAddress := d.GetOk(dsElasticIPAttrIPAddress)
-	if !byElasticIPID && !byElasticIPAddress {
+	elasticIPID, searchByElasticIPID := d.GetOk(dsElasticIPAttrID)
+	elasticIPAddress, searchByElasticIPAddress := d.GetOk(dsElasticIPAttrIPAddress)
+	elasticIPLabels, searchByElasticIPLabels := d.GetOk(dsElasticIPAttrLabels)
+	if !searchByElasticIPID && !searchByElasticIPAddress && !searchByElasticIPLabels {
 		return diag.Errorf(
-			"either %s or %s must be specified",
+			"one of %s, %s or %s must be specified",
 			dsElasticIPAttrIPAddress,
 			dsElasticIPAttrID,
+			dsElasticIPAttrLabels,
 		)
 	}
 
-	elasticIP, err := client.FindElasticIP(
-		ctx,
-		zone, func() string {
-			if byElasticIPID {
-				return elasticIPID.(string)
+	// search by address by default
+	filterElasticIP := func(eip *egoscale.ElasticIP) bool {
+		if eip.IPAddress.String() == elasticIPAddress {
+			return true
+		}
+		return false
+	}
+
+	if searchByElasticIPID {
+		filterElasticIP = func(eip *egoscale.ElasticIP) bool {
+			if *eip.ID == elasticIPID {
+				return true
+			}
+			return false
+		}
+	}
+
+	if searchByElasticIPLabels {
+		searchLabels := make(map[string]string)
+		for k, v := range elasticIPLabels.(map[string]interface{}) {
+			searchLabels[k] = v.(string)
+		}
+
+		filterElasticIP = func(eip *egoscale.ElasticIP) bool {
+			if eip.Labels == nil {
+				return false
+			}
+
+			for searchKey, searchValue := range searchLabels {
+				if foundValue, ok := (*eip.Labels)[searchKey]; ok && foundValue == searchValue {
+					continue
+				}
 			}
-			return elasticIPAddress.(string)
-		}(),
-	)
+
+			return true
+		}
+	}
+
+	elasticIPs, err := client.ListElasticIPs(ctx, zone)
 	if err != nil {
 		return diag.FromErr(err)
 	}
 
+	var elasticIP *egoscale.ElasticIP
+	for _, eip := range elasticIPs {
+		if filterElasticIP(eip) {
+			elasticIP = eip
+			break
+		}
+	}
+
+	if elasticIP == nil {
+		return diag.FromErr(fmt.Errorf("Unable to find matching ElasticIP"))
+	}
```

Comment:
We should discuss internally this part.
In my view, it could be way better to have this logic on the public-API side.